File: benchmarks/compare/raptor/go/scanresults.go
package raptor

// Connection is one round's connection to a stop: either a timetable
// connection [trip, startIndex, endIndex] or a transfer.
type Connection struct {
	// Transfer is non-nil for a transfer connection.
	Transfer *Transfer

	// The following are valid only when Transfer == nil (timetable leg).
	Trip       *Trip
	StartIndex int
	EndIndex   int
}

// IsTransfer reports whether this is a transfer connection (mirrors isTransfer
// which checks for an `origin` field).
func (c *Connection) IsTransfer() bool {
	return c.Transfer != nil
}

// Arrivals maps a stop to its best arrival time.
type Arrivals = map[StopID]Time

// RoundConnections maps a round number (k) to a Connection, preserving the
// rounds in numeric ascending order via sorted iteration in JourneyFactory.
type RoundConnections struct {
	byRound map[int]*Connection
	rounds  []int // insertion order; rounds are added strictly increasing here
}

func newRoundConnections() *RoundConnections {
	return &RoundConnections{byRound: map[int]*Connection{}}
}

func (r *RoundConnections) set(k int, c *Connection) {
	if _, ok := r.byRound[k]; !ok {
		r.rounds = append(r.rounds, k)
	}
	r.byRound[k] = c
}

func (r *RoundConnections) get(k int) *Connection { return r.byRound[k] }
func (r *RoundConnections) len() int              { return len(r.rounds) }

// ConnectionIndex maps a stop to its per-round connections.
//
// Keys (stop IDs) preserve insertion order to match JS object iteration in
// getFoundStations / GraphResults.getPaths / StringResults.add.
type ConnectionIndex struct {
	stops *OrderedMap[*RoundConnections]
}

func newConnectionIndex() *ConnectionIndex {
	return &ConnectionIndex{stops: NewOrderedMap[*RoundConnections]()}
}

// Stops returns stop IDs in insertion order.
func (ci *ConnectionIndex) Stops() []StopID { return ci.stops.Keys() }

// Round returns the per-round connections for a stop (may be empty).
func (ci *ConnectionIndex) Round(stop StopID) *RoundConnections {
	r, _ := ci.stops.Get(stop)
	return r
}

// HasResults reports whether a stop has at least one connection.
func (ci *ConnectionIndex) HasResults(stop StopID) bool {
	r, ok := ci.stops.Get(stop)
	return ok && r.len() > 0
}

// ScanResults accumulates per-round arrivals and connections during a scan.
type ScanResults struct {
	k            int
	bestArrivals Arrivals
	// kArrivals indexed by round; index 0 holds the origin arrivals. Each round
	// uses an OrderedMap so getMarkedStops returns insertion order.
	kArrivals   []*OrderedMap[Time]
	connections *ConnectionIndex
}

func (r *ScanResults) addRound() {
	r.k++
	r.kArrivals = append(r.kArrivals, NewOrderedMap[Time]())
}

func (r *ScanResults) previousArrival(stop StopID) (Time, bool) {
	return r.kArrivals[r.k-1].Get(stop)
}

func (r *ScanResults) setTrip(trip *Trip, startIndex, endIndex int, interchange Time) {
	time := trip.StopTimes[endIndex].ArrivalTime + interchange
	stop := trip.StopTimes[endIndex].Stop

	r.kArrivals[r.k].Set(stop, time)
	r.bestArrivals[stop] = time
	r.connections.Round(stop).set(r.k, &Connection{Trip: trip, StartIndex: startIndex, EndIndex: endIndex})
}

func (r *ScanResults) setTransfer(transfer *Transfer, time Time) {
	stop := transfer.Destination

	r.kArrivals[r.k].Set(stop, time)
	r.bestArrivals[stop] = time
	r.connections.Round(stop).set(r.k, &Connection{Transfer: transfer})
}

func (r *ScanResults) bestArrival(stop StopID) Time {
	return r.bestArrivals[stop]
}

func (r *ScanResults) getMarkedStops() []StopID {
	return r.kArrivals[r.k].Keys()
}

func (r *ScanResults) finalize() (*ConnectionIndex, Arrivals) {
	return r.connections, r.bestArrivals
}

// ScanResultsFactory builds a fresh ScanResults for a scan over a fixed set of
// stops.
type ScanResultsFactory struct {
	stops []StopID
}

// NewScanResultsFactory constructs the factory over the given stop universe.
func NewScanResultsFactory(stops []StopID) *ScanResultsFactory {
	return &ScanResultsFactory{stops: stops}
}

// Create initialises arrivals to the origin departure times (or +infinity) and
// an empty per-stop connection index.
func (f *ScanResultsFactory) Create(origins StopTimes) *ScanResults {
	bestArrivals := make(Arrivals, len(f.stops))
	round0 := NewOrderedMap[Time]()
	connections := newConnectionIndex()

	for _, stop := range f.stops {
		v := MaxSafeInteger
		if t, ok := origins.Get(stop); ok && t != 0 {
			v = t
		}
		bestArrivals[stop] = v
		round0.Set(stop, v)
		connections.stops.Set(stop, newRoundConnections())
	}

	return &ScanResults{
		bestArrivals: bestArrivals,
		kArrivals:    []*OrderedMap[Time]{round0},
		connections:  connections,
	}
}

File: benchmarks/compare/raptor/rust/src/scan_results.rs
//! ScanResults + ScanResultsFactory (port of src/raptor/ScanResults.ts).

use std::collections::BTreeMap;
use std::rc::Rc;

use indexmap::IndexMap;

use crate::gtfs::{StopId, Time, Transfer, Trip, MAX_SAFE_INTEGER};

/// A kConnection entry: either a boarded trip leg `[trip, start, end]` or a transfer.
#[derive(Debug, Clone)]
pub enum Connection {
    Trip(Rc<Trip>, usize, usize),
    Transfer(Transfer),
}

/// best arrival time per stop. Insertion-ordered (load-bearing for `getFoundStations`).
pub type Arrivals = IndexMap<StopId, Time>;

/// kConnections[stop][k]. Outer map preserves stop insertion order; inner map is
/// keyed by round number `k` and iterates **numeric ascending** (contract #1).
pub type ConnectionIndex = IndexMap<StopId, BTreeMap<usize, Connection>>;

/// kArrivals[k] = arrivals for round k; per-round map preserves insertion order
/// (the source of `getMarkedStops` ordering).
type RoundArrivals = IndexMap<StopId, Time>;

pub struct ScanResults {
    k: usize,
    best_arrivals: Arrivals,
    k_arrivals: Vec<RoundArrivals>,
    k_connections: ConnectionIndex,
}

impl ScanResults {
    pub fn add_round(&mut self) {
        self.k += 1;
        // kArrivals[++k] = {}. Indices 0..=k must exist.
        while self.k_arrivals.len() <= self.k {
            self.k_arrivals.push(IndexMap::new());
        }
    }

    pub fn previous_arrival(&self, stop_pi: &str) -> Option<Time> {
        self.k_arrivals[self.k - 1].get(stop_pi).copied()
    }

    pub fn set_trip(
        &mut self,
        trip: &Rc<Trip>,
        start_index: usize,
        end_index: usize,
        interchange: Time,
    ) {
        let time = trip.stop_times[end_index].arrival_time + interchange;
        let stop_pi = trip.stop_times[end_index].stop.clone();

        self.k_arrivals[self.k].insert(stop_pi.clone(), time);
        self.best_arrivals.insert(stop_pi.clone(), time);
        self.k_connections
            .entry(stop_pi)
            .or_default()
            .insert(self.k, Connection::Trip(Rc::clone(trip), start_index, end_index));
    }

    pub fn set_transfer(&mut self, transfer: &Transfer, time: Time) {
        let stop_pi = transfer.destination.clone();

        self.k_arrivals[self.k].insert(stop_pi.clone(), time);
        self.best_arrivals.insert(stop_pi.clone(), time);
        self.k_connections
            .entry(stop_pi)
            .or_default()
            .insert(self.k, Connection::Transfer(transfer.clone()));
    }

    pub fn best_arrival(&self, stop_pi: &str) -> Time {
        // JS reads bestArrivals[stopPi]; every relevant stop is seeded by the factory.
        *self.best_arrivals.get(stop_pi).unwrap_or(&MAX_SAFE_INTEGER)
    }

    pub fn get_marked_stops(&self) -> Vec<StopId> {
        self.k_arrivals[self.k].keys().cloned().collect()
    }

    pub fn finalize(self) -> (ConnectionIndex, Arrivals) {
        (self.k_connections, self.best_arrivals)
    }
}

pub struct ScanResultsFactory {
    /// `Object.keys(usefulTransfers)` — the seeded stops, in insertion order.
    stops: Vec<StopId>,
}

impl ScanResultsFactory {
    pub fn new(stops: Vec<StopId>) -> Self {
        ScanResultsFactory { stops }
    }

    pub fn create(&self, origins: &IndexMap<StopId, Time>) -> ScanResults {
        let mut best_arrivals: Arrivals = IndexMap::new();
        let mut round0: RoundArrivals = IndexMap::new();
        let mut k_connections: ConnectionIndex = IndexMap::new();

        for stop in &self.stops {
            let seed = origins.get(stop).copied().unwrap_or(MAX_SAFE_INTEGER);
            best_arrivals.insert(stop.clone(), seed);
            round0.insert(stop.clone(), seed);
            k_connections.insert(stop.clone(), BTreeMap::new());
        }

        ScanResults {
            k: 0,
            best_arrivals,
            k_arrivals: vec![round0],
            k_connections,
        }
    }
}
